Add support for fullscreenchange event to exit full screen mode

zagrim opened the pull request:
Closes #1967.

ccrisan commented:
This was actually more of a feature than a bug: having one single camera frame displayed while *not* in full-screen is no longer possible if this PR is merged.

lhartmann commented:
If single-camera-view is considered a feature, then having to go full-screen and back feels clunky.

Youtube has "theater" and "full-screen" modes separate. Maybe something similar then?

zagrim commented:
My dev setup had only one camera so I didn't even think of that, @ccrisan, but yeah, that's true :(

ccrisan commented:
> If single-camera-view is considered a feature, then having to go full-screen and back feels clunky.

Yes, i agree it's not the best experience one may have, but completely removing it isn't a solution, in my opinion.

MichaIng commented:
@zagrim 
Do you find time to work on a theatre mode implementation to have a single camera frame without being in fullscreen mode? That sounds like a very good solution to me, so exiting fullscreen can then revert to the previous layout.

zagrim commented:
I have some further changes locally, also that theatre mode / single-camera support, but looking at it now it seems that I ended up with a lack of ideas for how to make it behave in certain scenarios. There's quite a lot of opinions on the related issue, but I didn't find any commonly accepted way for the UI to behave, not to mention taking mobile use into account.
On top of that there seems to be a bug somewhere since with quick testing I twice arrived at a full-screen size UI while the browser was not full-screen (and the window not even maximised). So it would need some work.

I changed work projects at the start of 2021 and again after summer, and have not really had any extra motivation to try to finish this - partly also due to this project having pretty much unmaintained for the past year. Now with that last reason out of the window, at least, I might try to find time for this. But since this is luckily just UI stuff, any of the ongoing Python / build stuff doesn't break it at least.

MichaIng commented:
As you may have recognised, we moved the project into a new GitHub organisation with more and new members taking case of it now, most importantly migrating it to Python 3. Just as a motivation for you, considering that PRs are now reviewed and in case merged faster 🙂.

Feel free to rebase/recreate with your current WIP, so we can test and develop together.

zagrim commented:
Ok, I think I got things sorted out with my fork and updated the changes after some tweaks and testing.

One thing that is currently missing is applying translation to the added camera top bar button tooltip. I'd like to know @jmichault if there's some automation I should know to use to update the translation files with the added string or how is this supposed to go?

MichaIng commented:
Probably it makes sense to have some documentation about how to add and update translations added to the Wiki, or as part of a first `CONTRIBUTING.md` draft 🙂. I wouldn't see it as mandatory to add the translation here now, but of course, it wouldn't hurt.

MichaIng commented on motioneye/static/img/camera-top-buttons.svg:
```diff
@@ -171,23 +171,17 @@
-       d="m 212,1026.3622 10,6 -10,6 z"
-       id="path3819-3"
-       inkscape:connector-curvature="0"
-       sodipodi:nodetypes="cccc" />
```
You removed the time-lapse button 😉. Our of curiosity, since I guess you used inkscape to edit this SVG, are there any issues to do so with a minified SVG or does this cause havoc in inkscape? https://github.com/motioneye-project/motioneye/blob/imgbot/motioneye/static/img/camera-top-buttons.svg

Otherwise we might need to keep the original versions available separately for editing, minifying them to frontend assets in a second step.

zagrim replied:
Oops, need to fix that. :flushed:  I guess time-lapse was added after I started my branch...

zagrim replied:
Fixed. I didn't notice that Inkscape would've had any issues with the minification. AFAIK all the tags with namespace of `inkscape` or `sodipodi` are non-essential as long as only the actual image data is relevant (that is, no-one wants to store their preferred workspace settings and stuff in the file).

MichaIng replied:
> Fixed. I didn't notice that Inkscape would've had any issues with the minification.

Great, as I thought/hoped.

zagrim commented:
> I wouldn't see it as mandatory to add the translation here now, but of course, it wouldn't hurt.

As with documentation, if you leave it for later, you never end up doing it :smile:  But I'll wait for instructions on how to go on with translating, since the tool being used is totally foreign to me.

BTW, while I was wondering what to do with that, I also noted that there were some other untranslated simple strings nearby. Maybe I should start my journey learning Esperanto with those, too...

MichaIng commented:
> BTW, while I was wondering what to do with that, I also noted that there were some other untranslated simple strings nearby. Maybe I should start my journey learning Esperanto with those, too...

Indeed, hence the risk about this translation never being done is probably not so big, since we anyway need to go through all strings.

If you are interested in learning Esperanto anyway, that is awesome, but generally I vote for switching the main language to English, so contributors do not need to learn Esperanto first to be able to check/update/add translations for their language. Esperanto is an easy plan language, designed to be learned quickly, but in fact international communication works on English and possible translation contributors are discouraged if they need to learn the language first, even if it's done in 10 hours or so 😉.

zagrim commented:
> If you are interested in learning Esperanto anyway, that is awesome, but generally I vote for switching the main language to English, so contributors do not need to learn Esperanto first to be able to check/update/add translations for their language.

I was trying to be optimistic :slightly_smiling_face:  I'd rather have English as the base, too. I suspect the easiness of learning depends not only on the individual's capabilities but also the languages one already has learned. At least on the surface Esperanto resembles no languages I am able to converse with, so my guess is it'd take a lot more for me than, say, for someone fluent with some Central-European languages. But that's enough of that off-topic.

MichaIng commented:
I tested it, and works quite well. Was a hassle to get the SVG out of cache so that the new 280 wide one is used 😄.

Aside of the quirk you mentioned that when starting full screen from single cam view, that both buttons return to single cam view: As of the current logic of the buttons, meaning both, entering and exiting a view, this behaviour actually makes sense, but:
- When returning via single cam view button, the camera image is left in same size as it was in full screen, so it is then bigger than the screen.
- When entering full screen from multi cam view, both buttons return to multi cam view, which is unexpected when hitting the single cam view button, of course.

Generally, as mentioned in the issue, I think it is best to not have the same button for entering and exiting the mode. Probably it's best to add a multi cam view button and having always the other two mode buttons shown. So every button always means "enter this view" and exiting the view us always done explicitly to one of the other two via buttons, and only from full screen, the remembered one is used when hitting ESC/F11/browser shortcut? With three views otherwise it always remains a question what the same button means, and also what "exiting view" means. But if the above two quirks can be addressed otherwise, consistently returning to the previous view and having the camera frame size adjusted, it is fine for me both ways.

zagrim commented:
> When returning via single cam view button, the camera image is left in same size as it was in full screen, so it is then bigger than the screen.

This was the one I spotted on Saturday when checking the status of my changes, but I couldn't anymore replicate it on Sunday. Thanks for documenting it (and the other one which I didn't spot either for some reason, although I tried to go through all combinations) so I can debug.

> So every button always means "enter this view" and exiting the view us always done explicitly to one of the other two via buttons, and only from full screen, the remembered one is used when hitting ESC/F11/browser shortcut?

I agree that this might be the best approach here :+1 Shouldn't be too difficult to implement.

jmichault commented:
> 



> Ok, I think I got things sorted out with my fork and updated the changes after some tweaks and testing.
> 
> One thing that is currently missing is applying translation to the added camera top bar button tooltip. I'd like to know @jmichault if there's some automation I should know to use to update the translation files with the added string or how is this supposed to go?

* To update translations, run `make` in motioneye directory.
* Then you can enhance translations by editing the po files for the languages you are able to translate (they are named  locale/xx/LC_MESSAGES/motioneye.js.po for the js files), with a simple text editor, or better with dedicated software like poedit or virtaal.
* finally, re-run `make` to take your changes in translations into account

zagrim commented:
> When returning via single cam view button, the camera image is left in same size as it was in full screen, so it is then bigger than the screen.

This seems to be related to how `updateLayout` function works. I've not been able to get to the bottom of it, but it might be due to how it calculates the maximum camera frame ratio, which is done by looking at the actual height and width of corresponding DOM elements (got some wacky figures for that when the layout breaks). Anyway, when layout is updated during a transition from one view to another things just can go wrong.

I'm not sure if the camera frame element caching that is being used there was implemented just because of this, but seems like it is not enough, and OTOH disabling that also does not fix things. So I still need to debug this issue some more. A dirty fix might be to increase the usage of delayed calls to `updateLayout` (which has been there already in two places), but that feels bad. But still, since there are plenty of factors affecting the camera frame sizes and the UI doesn't even otherwise know the height/width ratio of each frame it might prove the easiest way.

zagrim commented:
Ok, I might have been able to find a working solution for the layout issue, too. When frame aspect ratios are stored based on the size the frame has when it gets refreshed, and using those values when updating the layout, at least one potential reason for layout breakage is eliminated, and now I cannot anymore reproduce odd layouts.

zagrim commented:
I also ran update for translations (as I modified one old and added two new strings). While doing this I noticed that while Makefile was moved up one directory the process started looking up translable strings also from `build/`. I'm not quite sure if the fix I included is ok, at least now strings are again searched only under `motioneye/`, but the source references in .po files still changed.

MichaIng commented on motioneye/locale/de/LC_MESSAGES/motioneye.js.po:
```diff
@@ -17,698 +17,715 @@ msgstr ""
-msgid "plena ekrano kamerao"
-msgstr "Vollbildkamera"
+msgid "montru ĉi tiun fotilon plenekranan"
+msgstr "Zeigen Sie diese Kamera Vollzeit"
```
```suggestion
msgstr "Diese Kamera im Vollbildmodus anzeigen"
```

MichaIng commented on motioneye/locale/de/LC_MESSAGES/motioneye.js.po:
```diff
@@ -17,698 +17,715 @@ msgstr ""
+#: motioneye/static/js/main.js:4706
+#, fuzzy
+msgid "montri ĉiujn fotilojn"
+msgstr "Alle Kameras anzeigen."
```
```suggestion
msgstr "Alle Kameras anzeigen"
```

zagrim replied:
Makes me wonder why the machine translation added the dot, as well as why it sometimes capitalises the first letter and sometimes not.

MichaIng commented on motioneye/locale/de/LC_MESSAGES/motioneye.js.po:
```diff
@@ -17,698 +17,715 @@ msgstr ""
+#: motioneye/static/js/main.js:4707
+#, fuzzy
+msgid "montru nur ĉi tiun fotilon"
+msgstr "Zeigen Sie nur diese Kamera"
```
```suggestion
msgstr "Nur diese Kamera anzeigen"
```

MichaIng commented:
Thanks for fixing the language build and updating the strings. I did some suggestions for the German language. Not sure how much space these texts actually have (to look nice), they could be shortened.

zagrim commented:
> Not sure how much space these texts actually have (to look nice), they could be shortened.

The text is shown as a tooltip on mouse hover so there's enough space.

MichaIng commented on motioneye/static/js/motioneye.de.json:
```diff
@@ -117,7 +117,9 @@
 ,"Bildoj prenitaj de ": "Bilder von"
 ,"Filmoj registritaj de ": "Filme aufgenommen von"
-,"plena ekrano kamerao": "Vollbildkamera"
+,"montru ĉi tiun fotilon plenekranan": "Zeigen Sie diese Kamera Vollzeit"
```
```suggestion
,"montru ĉi tiun fotilon plenekranan": "Diese Kamera im Vollbildmodus anzeigen"
```
Changed in sources but not in JSON yet. A re`make` would do as well, but here the quick manual update 🙂.

MichaIng commented on motioneye/static/js/motioneye.de.json:
```diff
@@ -117,7 +117,9 @@
 ,"Filmoj registritaj de ": "Filme aufgenommen von"
-,"plena ekrano kamerao": "Vollbildkamera"
+,"montru ĉi tiun fotilon plenekranan": "Zeigen Sie diese Kamera Vollzeit"
+,"montri ĉiujn fotilojn": "Alle Kameras anzeigen."
```
```suggestion
,"montri ĉiujn fotilojn": "Alle Kameras anzeigen"
```

MichaIng commented on motioneye/static/js/motioneye.de.json:
```diff
@@ -117,7 +117,9 @@
-,"plena ekrano kamerao": "Vollbildkamera"
+,"montru ĉi tiun fotilon plenekranan": "Zeigen Sie diese Kamera Vollzeit"
+,"montri ĉiujn fotilojn": "Alle Kameras anzeigen."
+,"montru nur ĉi tiun fotilon": "Zeigen Sie nur diese Kamera"
```
```suggestion
,"montru nur ĉi tiun fotilon": "Nur diese Kamera anzeigen"
```

MichaIng commented on motioneye/static/css/main.css:
```diff
@@ -82,6 +82,9 @@ div.header {
 div.header.full-screen {
     height: 0;
 }
+div.header.single-cam {
+    height: 0;
+}
```
```suggestion
div.header.full-screen,
div.header.single-cam {
    height: 0;
}
```

MichaIng commented on motioneye/static/css/main.css:
```diff
@@ -102,6 +105,9 @@ div.footer {
 div.footer.full-screen {
     height: 0;
 }
+div.footer.single-cam {
+    height: 0;
+}
```
```suggestion
```
Redundant when merged with header styles into a single block.

MichaIng commented on motioneye/static/css/main.css:
```diff
@@ -132,6 +138,10 @@ div.page-container.full-screen {
     padding-top: 0;
     padding-bottom: 0;
 }
+div.page-container.single-cam {
+    padding-top: 0;
+    padding-bottom: 0;
+}
```
```css
div.page-container.full-screen,
div.page-container.single-cam {
    padding-top: 0;
    padding-bottom: 0;
}
```
Same as above, merging blocks with same content.

MichaIng commented:
Tested, and it works very well, no visual glitches anymore. Also the code looks very good, excellent work 👍. Apart of the German language update (alternatively a `make` + commit would update them as well) and the CSS block merges, ready to go from my end.

MichaIng commented:
Suggestions added, merging this now.
